### PR TITLE
Molecular alteration CSV for Zenodo upload

### PR DESCRIPTION
**Pull request by sjspielman**
Part of #1692 

This PR _begins_ (code needs help/feedback during review!) the process of getting a molecular alterations csv. At this point, I decided to file for feedback since my brain keeps getting twisted thinking of the right way to join back with DNA and RNA ids for the ambiguous samples.

The new code involves:
- A script `tables/tabulate-molecular-alterations.R` to generate this CSV file
- A script `tables/util/functions-molecular-alterations.R` with some extremely context-specific functions, as indicated in comments in that file.

The `tables/tabulate-molecular-alterations.R` script takes the following steps until we get to:
1. Process maf, cnv, fusion data
    - mirroring the approach in https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/627ec427ad0a8d9d913e614c9db50546c56d8283/analyses/oncoprint-landscape/01-map-to-sample_id.R.
2. Combine those processed df's into a big alteration df
3. Filter to genes present in the manuscript oncoplots, and create a presence/absence (with "Y"/"N") data frame for whether each gene/alteration is present in each of the 1074 tumors.
4. Map back to biospecimen ids for export 🥴  

Step 4 is either fairly straightforward, or very not! 
- I have created `export_identifiable_ids` which is an export-ready wide data frame for _non-ambiugous_ ids that have 1:1 DNA:RNA. 
- But, I still need a solid strategy for dealing with the non 1:1 mappings. There's two types of problematic mappings:
  - Where there are >1 RNA or DNA experiments, so we can't directly line up a DNA with RNA. This is a 1:many situation, like we discussed.
  - Where there are >1 of a given modality - eg 2 RNAs, or even a case with 6 DNAs!. This is, well, just a plain "many."

One potential approach that I was thinking of is to just add these `sample_id`s and alterations into the main export df (called `export_identifiable_ids` at the moment), _but_ keep their biospecimen id values as `NA` . Then, we could have a separate file to indicate what those ids might be, with columns: 
- `sample_id`
- `Kids_First_Biospecimen_ID`
- `experimental_strategy`
I don't think this approach will make a downstream user's life easier, but I also think they're (we're?) just in for a hard time anyways with these samples...

### CODE REVIEW

**Comment by sjspielman** [comment]
Requested @jaclyn-taroni for review here mostly for purposes of strategizing ambiguous ids as needed!

**Comment by sjspielman** [comment]
This closes #1726.

File: tables/tabulate-molecular-alterations.R
Comment (jashapiro) [inline tables/tabulate-molecular-alterations.R:167]: Each of these functions is used only once, so I'm not sure the value of having them separated as functions, especially in a separate file. At the least, I would like to see here something about what this preparation is: what are the changes that are being made by these functions?

File: tables/tabulate-molecular-alterations.R
Comment (jashapiro) [inline tables/tabulate-molecular-alterations.R:217]: Any reason not to use T/F here?
```suggestion
  dplyr::mutate(present = TRUE) %>% 
  # Fill in all combinatins of gene/variant for each sample, 
  #  filling in FALSE for present for any new rows 
  tidyr::complete(sample_id, Hugo_Symbol, Variant_Classification, 
                  fill = list(present = FALSE)) %>%
```

File: tables/tabulate-molecular-alterations.R
Comment (jashapiro) [inline tables/tabulate-molecular-alterations.R:233]: If we don't filter here, we will end up with multiple rows for each `sample_id` where there may be more than one DNA or RNA sample after joining. If we then pass this through a `group_by()`, we should be able to collapse DNA & RNA columns with a `paste` or similar, keeping everything else the same. 

Does that make sense as a strategy?

That at least should work for multiple modalities... I'm not sure what the ambiguous samples really look like.

**Review by jashapiro — COMMENTED** [review]
Some initial comments here, the last of which is a potential strategy comment.

I'm confused by the difference between ambiguous and multi-mapping. From your description they both sound like multimapping, just of degree? Either way, I think making the DNA/RNA columns be semicolon-separated lists of ids is the most straightforward, and accomplishing that with an appropriate group_by/summarize seems the simplest approach. But I could be missing something major here!

File: tables/tabulate-molecular-alterations.R
Comment (sjspielman) [reply tables/tabulate-molecular-alterations.R:167]: The value of having them in separate functions is mostly for my brain, less clutter during coding. All of that code was all originally in this script, but as the script got bigger and bigger....

Can definitely de-function them in the end, and I'll add some more comments here.

**Comment by sjspielman** [comment]
Hm I'd just like to note, for the record, that this branch name, which I personally named, is....not how my name is spelled.

**Comment by sjspielman** [comment]
>I'm confused by the difference between ambiguous and multi-mapping. From your description they both sound like multimapping, just of degree? 

Yes, you've got it! Both are ambiguous situations, but I should have been clear to say that one flavor is ambiguity for mapping RNA-DNA (1:many situation), and the other flavor is, for example, which DNA id (if there are several) is being referred to. The latter is not necessary 1:many.

> Either way, I think making the DNA/RNA columns be semicolon-separated lists of ids is the most straightforward, and accomplishing that with an appropriate group_by/summarize seems the simplest approach. But I could be missing something major here!

I don't disagree at all, but I have in my notes from yesterday's chat about this that we decided to avoid that strategy and instead have a separate column to indicate if the given _single pair_ of RNA/DNA ids is a 1:1 or 1:many situation. 
Maybe I'll go for a `why-not-both.gif` situation - have semi-colon separated ids _and_ a T/F column to indicate "ambiguous mapping".

**Comment by sjspielman** [comment]
> Maybe I'll go for a `why-not-both.gif` situation - have semi-colon separated ids and a T/F column to indicate "ambiguous mapping".

As a move towards next steps, I went ahead and did this in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/1725/commits/42d7ee84feb110fdced906871471be3d57a50f7a. The output CSV is manageable at 8MB without compression, but can compress if we want a tiny file!

File: tables/tabulate-molecular-alterations.R
Comment (jashapiro) [inline tables/tabulate-molecular-alterations.R:369]: I'd leave out the space: easier to unparse later.

Also, just in case, adding unique and sort here: 
```suggestion
  dplyr::summarize(
    Kids_First_Biospecimen_ID = paste(
      sort(unique(Kids_First_Biospecimen_ID)),
      collapse=";")
  ) %>%
```

File: tables/tabulate-molecular-alterations.R
Comment (jashapiro) [inline tables/tabulate-molecular-alterations.R:371]:
```suggestion
  # bring back the other variables from `tumor_sample_ids_df`
```

File: tables/tabulate-molecular-alterations.R
Comment (jashapiro) [reply tables/tabulate-molecular-alterations.R:369]: Further thought: if you don't filter to just ambiguous, you can put that indicator here as well, and save having to have two flows, I think. 

```suggestion
    dplyr::summarize(
    Kids_First_Biospecimen_ID = paste(
      sort(unique(Kids_First_Biospecimen_ID)),
      collapse=";"),
    ambiguous_biospecimen_mapping = dplyr::n()>1
  ) %>%
```

File: tables/tabulate-molecular-alterations.R
Comment (jashapiro) [inline tables/tabulate-molecular-alterations.R:400]:
```suggestion
if (length(unique(tumor_sample_ids_df$sample_id)) != openpbta_n_tumors) {
```

File: tables/tabulate-molecular-alterations.R
Comment (jashapiro) [inline tables/tabulate-molecular-alterations.R:417]: There may be two values here, right? One each for DNA & RNA? do we need to account for that in the spread?

**Review by jashapiro — COMMENTED** [review]
I think we are getting there. I had a suggestion to simplify a bit more, but I also had a question about the final output for the ambiguous indicator... I feel like it may be easiest to just leave that to the very end and add it if there is a semicolon in DNA or RNA columns, rather than trying to maintain it by other means through the transformations.

File: tables/tabulate-molecular-alterations.R
Comment (jaclyn-taroni) [inline tables/tabulate-molecular-alterations.R:237]: We want to include cell lines, too, per the response to reviewers and https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/1692#issue-1626351515 (adding emphasis here):

> A CSV file that contains a table of molecular alterations included in Figure 2 for all tumor **and cell line samples** in the PBTA dataset

Comment (jaclyn-taroni) [reply tables/tabulate-molecular-alterations.R:237]: I expect this also impacts your use of `openpbta_n_tumors` throughout. You could add `openpbta_n_cell_line` and then have the check be for the length of `openpbta_n_tumors + openpbta_n_cell_line`.

File: tables/tabulate-molecular-alterations.R
Comment (jaclyn-taroni) [inline tables/tabulate-molecular-alterations.R:284]: I would probably name the variable such that it's clear this is a _count_.

**Comment by sjspielman** [comment]
@jashapiro while working to continue cleaning this process up, I came across this neat situation which I did not anticipate, but glad I caught it:
Screenshot 2023-03-29 at 3 37 22 PM

The two RNA samples have "different" sexes - `Male` and `NA`. So, if we have these ids on the same row, then we have a conflict with this metadata variable. On one hand, we could defer to the non-`NA` value here and proceed with `Male`, but also I am wary about changing metadata here. But back to the first hand, we could document here that we use the non-`NA` sex if this conflict occurs.

And on a third hand, it seems that there are only 2 sample id's like this! For such a limited scenario, maybe just using the non-`NA` and explaining in `README` is indeed the way to go.
```
tumor_sample_ids_df %>% 
  dplyr::select(sample_id, germline_sex_estimate) %>% 
  dplyr::distinct() %>%
  dplyr::count(sample_id) %>%
  dplyr::filter(n>1)
# A tibble: 2 x 2
  sample_id     n
  <chr>     <int>
1 7316-161      2
2 7316-85       2
```

**Review by jaclyn-taroni — COMMENTED** [review]
Returning some initial comments. More importantly, **I was wrong yesterday!** I’ve re-reviewed the reviewer’s comments, and I don’t think this should be the presence or absence of alteration types (e.g., “Intron”) in the oncoprint. Instead, we’re being asked to include the _specific alterations_ in a table like so (this is a made-up example):

|  ids ... `histologies` data columns | TP53 |  BRAF |
|---|--|--|
| … |  p.A159P | None  |
| … | None |  KIAA1549--BRAF  |

The genes as columns will be the genes you have in `gene_list`. This might require a lot less wrangling for the fusions, for example (I very well might be wrong about that). 

In the case of _multiple alterations_, we can semi-colon separate them.

**Comment by sjspielman** [comment]
> For 1:1 mappings between assays, I would expect one row per `sample_id`.

Going back to update code accordingly to have 1 row where possible, one item occurred to me as a _potential_ reason to maybe not do that that I'd like to hear thoughts on! The review itself asks for much more metadata to be part of this export, including whether certain experimental modalities are available (column for `has_RNAseq`, `has_WGS`, etc.). These kinds of columns will end up doubling down on some ambiguity - e.g. sure WGS might be available, but which biospecimen is it? Having 1 row for each biospecimen locks down this particular metadata item, and I think it gives more clarity as to which biospecimen is each type of experiment. 

Either way, I'm working now towards 1-row-per either way, and this is something to think about!

**Comment by jaclyn-taroni** [comment]
@jashapiro I'm making edits to this and will tag you when it is ready for a look on

**Comment by jaclyn-taroni** [comment]
Okay, we now have a CSV file with the following columns:

| `sample_id` | `Kids_First_Biospecimen_ID_DNA` | `Kids_First_Biospecimen_ID_RNA` | `multiple_assays_within_type` | `sample_type` | `composition` | `germline_sex_estimate` | `broad_histology` | `cancer_group` | `ACVR1` | ... | `ZIC1` |
|-----------|---|---|---|---|---|---|---|---|---|---|---|

Where: 

* (Almost) each row is a unique `sample_id`-`composition` pair – more on the "almost" in a second.
* `Kids_First_Biospecimen_ID_DNA` are all DNA assay biospecimen IDs associated with the `sample_id`-`composition` pair in that row, separated by `; `.
* * `Kids_First_Biospecimen_ID_RNA` are all RNA assay biospecimen IDs associated with the `sample_id`-`composition` pair in that row, separated by `; `.
* `multiple_assays_within_type` indicates whether there is more than one DNA or more than one RNA assay for that `sample_id`-`composition` pair.
* A gene column will have all alterations found in at least one biospecimen for each assay type (RNA and DNA). For multiple alterations, they are separated by `; `. If no alterations were detected given the available assays, it says "None."

You can see how I dealt with the issue mentioned in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/1725#issuecomment-1489204244 here: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/1725/files#diff-7eed238487c7c49c705f8ab849980cf0183d136fd84054fc833cdbe656cc14e7R365-R375

One `sample_id`-`composition` pair has multiple cancer groups / molecular subtype, resulting in 1075 rows:

```r
   > histologies_df %>% filter(sample_id == "7316-3230") %>% select(sample_id, Kids_First_Biospecimen_ID, composition, broad_histology,cancer_group, molecular_subtype)
# A tibble: 4 × 6
  sample_id Kids_First_Biospecimen_ID composition  broad_histology                               cancer_group                  molecular_subtype          
  <chr>     <chr>                     <chr>        <chr>                                         <chr>                         <chr>                      
1 7316-3230 BS_5968GBGT               Solid Tissue Diffuse astrocytic and oligodendroglial tumor Diffuse midline glioma        DMG, H3 K28, TP53 loss     
2 7316-3230 BS_AF5D41PD               Solid Tissue Diffuse astrocytic and oligodendroglial tumor Diffuse midline glioma        DMG, H3 K28, TP53 loss     
3 7316-3230 BS_EE73VE7V               Solid Tissue Diffuse astrocytic and oligodendroglial tumor High-grade glioma astrocytoma HGG, H3 wildtype, TP53 loss
4 7316-3230 BS_HYKV2TH9               Solid Tissue Diffuse astrocytic and oligodendroglial tumor Diffuse midline glioma        DMG, H3 K28, TP53 loss   
```


The instance with `HGG, H3 wildtype, TP53 loss` reports a different primary site. Not really sure what to do about this issue here. cc: @jharenza

**Comment by jaclyn-taroni** [comment]
@jashapiro this is ready for a look. There an issue with one sample I noted here: https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/1725#issuecomment-1492444139

File: tables/tabulate-molecular-alterations.R
Comment (sjspielman) [inline tables/tabulate-molecular-alterations.R:245]: 👀 this is cute, new to me

File: tables/tabulate-molecular-alterations.R
Comment (sjspielman) [inline tables/tabulate-molecular-alterations.R:254]: I confirmed there are no `NA`s at this point in `final_alteration` 👍

**Review by sjspielman — COMMENTED** [review]
Thanks for tackling a lot of this while I was out!! I'm noticing some ids that may not have made it in the end, but still working out what's up. Pasting in here some preliminary investigating - 

```
# All the non-Normal ids
expected_sample_ids <- histologies_df %>% 
  dplyr::filter(sample_type != "Normal") %>% 
  dplyr::pull(sample_id) %>% 
  unique()
length(expected_sample_ids)
# [1] 1080

# What actually made it in the end?
final_sample_ids <- unique(final_alterations_df$sample_id)
length(final_sample_ids)
# [1] 1064

# Hm, those lengths are different. What's missing?
expected_sample_ids[!(expected_sample_ids %in% final_sample_ids)]
#  [1] "A14709" "A19649" "A16538" "A15170" "A17997" "A16405" "A19015"
#  [8] "A18777" "A15726" "A20091" "A16915" "A14565" "A14567" "A18552"
# [15] "A17096" "A12911"

histologies_df %>%
  dplyr::filter(sample_id %in% expected_sample_ids) %>%
  dplyr::select(sample_id, experimental_strategy, sample_type, composition) %>%
  head()
# #A tibble: 6 x 4
# sample_id experimental_strategy sample_type composition           
# <chr>     <chr>                 <chr>       <chr>                 
#  1 7316-2558 WGS                   Tumor       Solid Tissue          
#  2 7316-2753 WGS                   Normal      Peripheral Whole Blood
#  3 7316-2660 WGS                   Tumor       Solid Tissue          
#  4 7316-897  WGS                   Normal      Peripheral Whole Blood
#  5 7316-362  RNA-Seq               Tumor       Solid Tissue          
#  6 7316-3299 WGS                   Tumor       Solid Tissue 

# Hm, why/how is there a Normal coming up there?!
# ...because it's both...!
histologies_df %>%
  dplyr::filter(sample_id == "7316-2753") %>%
  dplyr::select(sample_id, experimental_strategy, sample_type, composition)
# A tibble: 3 x 4
#  sample_id experimental_strategy sample_type composition           
#   <chr>     <chr>                 <chr>       <chr>                 
# 1 7316-2753 WGS                   Normal      Peripheral Whole Blood
# 2 7316-2753 WGS                   Tumor       Solid Tissue          
# 3 7316-2753 RNA-Seq               Tumor       Solid Tissue    
```

**Comment by jaclyn-taroni** [comment]
Shooting from the hip – I suspect the missing sample IDs have something to do with the copy number alteration data because I believe all of those samples are assayed with WXS. We only have CNV data for WGS samples, if I recall correctly.

**Comment by jaclyn-taroni** [comment]
Now having looked, I'm reasonably sure it's the handling of `NA` germline sex estimate for WXS samples.

**Comment by jaclyn-taroni** [comment]
I made some changes in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/1725/commits/d13c5086ff6a8e85e0e6b8c82fa0f69e75f79466. Please take another look @sjspielman!

**Comment by jharenza** [comment]
>One sample_id-composition pair has multiple cancer groups / molecular subtype, resulting in 1075 rows:

   > histologies_df %>% filter(sample_id == "7316-3230") %>% select(sample_id, Kids_First_Biospecimen_ID, composition, broad_histology,cancer_group, molecular_subtype)
# A tibble: 4 × 6
  sample_id Kids_First_Biospecimen_ID composition  broad_histology                               cancer_group                  molecular_subtype          
  <chr>     <chr>                     <chr>        <chr>                                         <chr>                         <chr>                      
1 7316-3230 BS_5968GBGT               Solid Tissue Diffuse astrocytic and oligodendroglial tumor Diffuse midline glioma        DMG, H3 K28, TP53 loss     
2 7316-3230 BS_AF5D41PD               Solid Tissue Diffuse astrocytic and oligodendroglial tumor Diffuse midline glioma        DMG, H3 K28, TP53 loss     
3 7316-3230 BS_EE73VE7V               Solid Tissue Diffuse astrocytic and oligodendroglial tumor High-grade glioma astrocytoma HGG, H3 wildtype, TP53 loss
4 7316-3230 BS_HYKV2TH9               Solid Tissue Diffuse astrocytic and oligodendroglial tumor Diffuse midline glioma        DMG, H3 K28, TP53 loss   
The instance with HGG, H3 wildtype, TP53 loss reports a different primary site. Not really sure what to do about this issue here. cc: @jharenza

This was the one sample in which VarDict was the only algorithm which called H3K28M - it was definitely real, and prompted us to use clinical reports of H3 status (below), but this sample did not have a post-mortem clinical report and thus, it was decided we would not change this one to match the other samples taken at the same time (_I think_). I think it would be fine to merge all of the DNA for this sample into one field and use the combination of molecular alterations (eg if one sample has ABC and one has BCD, then the resulting alterations would be ABCD).



https://github.com/AlexsLemonade/OpenPBTA-analysis/blob/cf6328ebd7c3bdffa18b4bd6b8d856713a8414aa/analyses/molecular-subtyping-pathology/03-incorporate-clinical-feedback.Rmd#L35-L36

File: tables/tabulate-molecular-alterations.R
Comment (jharenza) [inline tables/tabulate-molecular-alterations.R:82]: Can we use the `HGVSp_Short` here?

File: tables/tabulate-molecular-alterations.R
Comment (jashapiro) [inline tables/tabulate-molecular-alterations.R:103]: Agree that short form makes sense here, but we probably also want to use the nucleotide change if there is no change in the protein, which I am doing here by grabbing the last character, which will be `=` if there is no change. 
```suggestion
      HGVSp_Short != "."  & 
        stringr::str_sub(HGVSp_Short, -1) != "=" ~ HGVSp_Short,
      HGVSc != "." ~ HGVSc,
```

**Review by jashapiro — COMMENTED** [review]
This looks good to me. I initially had the one comment about whether we want `; ` or `;` as the separator (I favor the latter), but then I did a little MAF diving and found something else that may be worth changing.

File: tables/tabulate-molecular-alterations.R
Comment (jaclyn-taroni) [reply tables/tabulate-molecular-alterations.R:29]: I find the `for` loop easier to read in this case (not always the case!)

**Comment by sjspielman** [comment]
Here is something we need to solidify - Pictured are the rows whose given `sample_id` appears in more than one row in the `final_alteration_df`. A single sample can be in multiple rows if there is a different composition associated, which seems fair to me. But, have a look at the `multiple_assays_within_type` column, and let's consider the two cases:
- `7316-1746`: Has `TRUE` in this column for _both_ associated rows, even though one of the compositions does not have more than 1 assay
- `7316-1769`: Has `FALSE` in this column for both associated rows. 

So, how do we want to consider if there is >1 assay? Is this at the level of `sample_id`, or _joint_ `sample_id` and `composition`. 


Screenshot 2023-04-03 at 3 22 21 PM

**Comment by sjspielman** [comment]
Also we need to have a specific look at one of those samples `"7316-3230"` which has, _excitingly_, different cancer groups, and therefore ends up as an "incorrectly duplicated" row.
Screenshot 2023-04-03 at 3 35 56 PM

From `histologies`:
Screenshot 2023-04-03 at 3 37 03 PM


Not really sure what to make of this one in general...

**Comment by jaclyn-taroni** [comment]
> Also we need to have a specific look at one of those samples `"7316-3230"` which has, _excitingly_, different cancer groups, and therefore ends up as an "incorrectly duplicated" row. Screenshot 2023-04-03 at 3 35 56 PM
> 
> From `histologies`: Screenshot 2023-04-03 at 3 37 03 PM
> 
> Not really sure what to make of this one in general...

Are you seeing this after the changes in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/1725/commits/b0f9d2881ef41954bf87644179d7ded4dc068a04?

**Comment by jaclyn-taroni** [comment]
> 

Thanks for catching this! Should be fixed in https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/1725/commits/ed6f28fbb11b089db138aabe51a6c3a605108dfd

**Comment by sjspielman** [comment]
Ah, I had missed https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/1725/commits/b0f9d2881ef41954bf87644179d7ded4dc068a04! :tada: it's not there, success :)

**Comment by sjspielman** [comment]
I can't approve this as it is "my PR", but I think we're there, at least as far as I can see!

**Review by jashapiro — APPROVED** [review]
> I can't approve this as it is "my PR", but I think we're there, at least as far as I can see!

I can!

**Pull request merged**
